File: create_token.rb
# Darren Baptiste darren@darrenbaptiste.com
#
# With thanks to https://github.com/jwt/ruby-jwt
# Hat tip to Dan Devine <https://stackoverflow.com/users/611955/dan-devine> for mentioning the need to add the 'typ' when using this gem

require 'jwt'

# ***** BEGIN CONFIGURATION == CHANGE THE ITEMS BELOW ********

private_key = '-----BEGIN PRIVATE KEY-----
ABCDEF123GHIJ456KLMN78900PQR1234STU56VWX789Z0
-----END PRIVATE KEY-----'

keyId = 'YOUR-10-DIGIT-APPLE-MUSIC-KEY-IDENTIFIER'
teamId = 'YOUR-10-DIGIT-APPLE-TEAM-ID'
hours_to_live = 24

# ***** END OF CONFIGURATION == ONLY CHANGE THE ITEMS ABOVE ********


time_now = Time.now.to_i
time_expired = Time.now.to_i + hours_to_live * 3600
algorithm = 'ES256'

headers = {
	'typ': 'JWT',
	'kid': keyId
}

payload = {
	'iss': teamId,
	'exp': time_expired,
	'iat': time_now
}

ecdsa_key = OpenSSL::PKey::EC.new private_key
ecdsa_key.check_key

token = JWT.encode payload, ecdsa_key, algorithm, header_fields=headers

puts token


# Uncomment the following section is for testing our output, including the generation of a curl script
#
# ecdsa_public = OpenSSL::PKey::EC.new ecdsa_key
# ecdsa_public.private_key = nil
# decoded_token = JWT.decode token, ecdsa_public, true, { algorithm: algorithm }
#
# puts decoded_token
#
# puts "curl -v -H \'Authorization: Bearer #{ token }\' \"https://api.music.apple.com/v1/catalog/ca/genres\" "

File: create_token.js
require('dotenv').config();
const jwt = require('jsonwebtoken');
const fs = require('fs');
const inquirer = require('inquirer');

const hours_to_live = 24
const header = {
	'typ': 'JWT',
	'kid': ''
}
const payload = {
	'iss': '',
	'exp': Math.floor((Date.now() / 1000) + hours_to_live * 3600),
	'iat': Math.floor(Date.now() / 1000)
}

//const privateKey = fs.readFileSync('private.key');
//const keyId = 'YOUR-10-DIGIT-APPLE-MUSIC-KEY-IDENTIFIER'
//const teamId = 'YOUR-10-DIGIT-APPLE-TEAM-ID'

let privateKey;

if(process.argv.length === 3){
    // arg[0] - private key
    // arg[1] - keyId
    // arg[2] - teamId
} else if(process.argv.length === 1 && process.env.TEAM_ID && process.env.KEY_ID){
    privateKey = fs.readFileSync(response.privateKeyFilename);
} else {
    // Use inquirer to collect information from user via command line
    inquirer
        .prompt([
            {
                type: 'input',
                message: 'Private key file name:',
                name: 'privateKeyFilename',
            },
            {
                type: 'input',
                message: 'Key ID:',
                name: 'kid',
            },
            {
                type: 'input',
                message: 'Team ID:',
                name: 'teamID',
            },
        ])
        .then((response) => {
            privateKey = fs.readFileSync(response.privateKeyFilename);
            payload.iss = response.teamId;
            header.kid = response.kid;
        })
}

// formed like:
// jwt.sign(payload, secretOrPrivateKey, [options, callback])
jwt.sign(payload, privateKey, { algorithm: 'RS256', header }, function(error, token) {
    console.log(token);
});